src/calc_integrate.jl

```diff
@@ -0,0 +1,325 @@
+    prec::Clong,
+)
+    return ccall(
+        @libarb(acb_calc_integrate),
+        Cint,
+        (
+            Ref{acb_struct},
+            Ptr{Cvoid},
+            Any,
+            Ref{acb_struct},
+            Ref{acb_struct},
+            Clong,
+            Ref{mag_struct},
+            Ptr{Cvoid},
+            Clong,
+        ),
+        res,
+        acb_calc_func,
+        param,
+        a,
+        b,
+        rel_goal,
+        abs_tol,
+        options,
+        prec,
+    )
+end
+
+"""
+    _integrate!(res, f, a, b; prec, rel_goal, abs_goal, opts)
+
+Internal function for integration. Assumes that `f!` is of the type
+`f!(out::AcbRef, inp::AcbRef; analytic::Bool, prec::Integer)` and
+that all other parameters are set to valid values.
+"""
+function _integrate!(
+    f!,
+    res::AcbLike,
+    a::AcbLike,
+    b::AcbLike;
+    prec::Integer = precision(res),
+    rel_goal::Integer = prec,
+    abs_tol::MagLike = set_ui_2exp!(Arblib.Mag(), one(UInt), -prec),
+    opts::Union{Ptr{Cvoid},calc_integrate_opt_struct} = C_NULL,
+)
+    status = calc_integrate!(
+        res,
+        _acb_calc_cfunc!(),
+        f!,
+        a,
+        b,
+        rel_goal,
+        abs_tol,
+        opts, # passing C_NULL uses the default options
+        prec,
+    )
+
+    # status:
+    # ARB_CALC_SUCCESS = 0
+    # ARB_CALC_NO_CONVERGENCE = 2
+    status == 2 && @warn "integration did not converge"
+
+    return res
+end
+
+"""
+    integrate!(f!, res::Acb, a::Number, b::Number;
+        check_analytic::Bool = false,
+        take_prec::Bool = false,
+        prec::Integer = precision(res),
+        rtol = 0.0,
+        atol = 2.0^-prec,
+        opts::Union{acb_calc_integrate_opt_struct, Ptr{Cvoid}} = C_NULL,
+    )
+
+Like [`integrate`](@ref), but make use of in-place operations. In
+particular, there are three differences from `integrate`:
+
+1. The function `f!` should be of the form `f!(y, x) = set!(y, f(x))`.
+   That is, it writes the return value of the integand `f(x)` in-place
+   into its first argument `y`. (The return value of `f!` is ignored).
+
+2. `res` is set to the result.
+
+3. The default precision is taken from `res` instead of from `a` and
+   `b`.
+"""
+function integrate!(
+    f!,
+    res::Union{Acb,AcbRef},
+    a::Union{Acb,AcbRef},
+    b::Union{Acb,AcbRef};
+    check_analytic::Bool = false,
+    take_prec::Bool = false,
+    prec::Integer = precision(res),
+    rtol = 0.0,
+    atol = set_ui_2exp!(Mag(), one(UInt), -prec),
+    opts::Union{Ptr{Cvoid},calc_integrate_opt_struct} = C_NULL,
+)
+    # rtol ≈ 2^(-rel_goal)
+    rel_goal = iszero(rtol) ? prec : max(-floor(Int, log2(atol)), 0)
+
+    if !check_analytic && !take_prec
+        g! = (inp, out; analytic, prec) -> f!(inp, out)
+    elseif !check_analytic && take_prec
+        g! = (inp, out; analytic, prec) -> f!(inp, out, prec = prec)
+    elseif check_analytic && !take_prec
+        g! = (inp, out; analytic, prec) -> f!(inp, out, analytic = analytic)
```

Comment:
change order of `inp`, `out` ;) I'd like to have these named `f!(res, x, ...)`, but that might be confusing with the `res` argument (though I believe perfectly valid).

I'm not sure how expensive defining (and compiling) those anonymous functions will be;
Another solution would be to define
```julia
struct CalcFunction{Analytic, TakePrec, F}
    func::F
    CalcFunction(f, inplace::Val{true}; check_analytic::Bool=false, take_prec=false) = 
        new{check_analytic, take_prec}(f)
    CalcFunction(f, inplace::Val{false}; check_analytic::Bool=false, take_prec=false) = 
        new{check_analytic, take_prec}(res, x) -> set!(res, f(x)))
end

(F::CalcFunction{false,false})(res, x; analytic, prec) = F.func(res, x)
(F::CalcFunction{true,false})(res, x; analytic, prec) = F.func(res, x, analytic=analytic)
....
```
and pass this guy directly as `param`. I'm not sure if this would actually be more efficient...
I'd be actually interested if deparametrizing `func` would actually change performance ;)

Reply:
So I did:
```julia
struct _CalcFunction{Analytic, Precision, F} <: Function
    func::F
    _CalcFunction(f; check_analytic::Bool=false, take_prec::Bool=false) = new{check_analytic, take_prec, typeof(f)}(f)
end

(F::_CalcFunction{false, false})(res::AcbLike, x::AcbLike; analytic, prec) =
    F.func(res, x)
(F::_CalcFunction{true, false})(res::AcbLike, x::AcbLike; analytic, prec) =
    F.func(res, x, analytic=analytic)
(F::_CalcFunction{false, true})(res::AcbLike, x::AcbLike; analytic, prec) =
    F.func(res, x, prec=prec)
(F::_CalcFunction{true, true})(res::AcbLike, x::AcbLike; analytic, prec) =
    F.func(res, x, analytic=analytic, prec=prec)

function integrate2!(
    f!,
    res::Union{Acb,AcbRef},
    a::Union{Acb,AcbRef},
    b::Union{Acb,AcbRef};
    check_analytic::Bool = false,
    take_prec::Bool = false,
    prec::Integer = precision(res),
    rtol = 0.0,
    atol = set_ui_2exp!(Mag(), one(UInt), -prec),
    opts::Union{Ptr{Cvoid},calc_integrate_opt_struct} = C_NULL,
)
    # rtol ≈ 2^(-rel_goal)
    rel_goal = iszero(rtol) ? prec : max(-floor(Int, log2(atol)), 0)

    g! = _CalcFunction(f!, check_analytic=check_analytic, take_prec=take_prec)

    return _integrate!(
        g!,
        res,
        a,
        b,
        prec = prec,
        rel_goal = rel_goal,
        abs_tol = convert(Mag, atol),
        opts = opts,
    )
end
```

benchmark:
```julia
using Arblib

using Test
using BenchmarkTools
using QuadGK

let prec = 256
    a = Acb(0, prec = prec)
    b = Acb(1, prec = prec)
    out = Acb(a)
    @btime Arblib.integrate!($(Arblib.sin!), $(out), $(a), $(b))
    @btime Arblib.integrate2!($(Arblib.sin!), $(out), $(a), $(b))
    setprecision(BigFloat, prec) do
        @btime quadgk($sin, $(big(0.0)), $(big(1.0)))
    end
end;
```
gives:
```
# first time:
  78.148 μs (142 allocations: 5.67 KiB)
  78.828 μs (142 allocations: 5.67 KiB)
  3.678 ms (17660 allocations: 805.05 KiB)
# second time
  78.237 μs (138 allocations: 5.55 KiB)
  78.685 μs (138 allocations: 5.55 KiB)
  3.709 ms (17660 allocations: 805.05 KiB)
```

There is just 2µs impact when `func::Any`; not bad :)

Reply:
in the same situation
```julia
@btime Arblib.integrate($sin, $a, $b)
```
takes
```
82.385 μs (223 allocations: 12.08 KiB)
```
this looks really great!

Reply:
Okay so it seems like we are good with the current version? I can change the arguments to `g!`. The mimic the ones for Arb at the moment, but it's probably better to follow Julia standards.

Reply:
I realised I had also just named them in the wrong direction...

Reply:
measuring the time to define an anonymous function:
```julia
let prec = 256
    a = Acb(0, prec = prec)
    b = Acb(1, prec = prec)
    out = Acb(a)
    @btime Arblib.integrate!($(Arblib.sin!), $(out), $(a), $(b))
    let g = (res, x; analytic, prec) -> Arblib.sin!(res, x)
        @btime Arblib._integrate!($g, $(out), $(a), $(b))
    end
    @btime Arblib.integrate2!($(Arblib.sin!), $(out), $(a), $(b))
```
```
  78.406 μs (138 allocations: 5.55 KiB)
  77.090 μs (138 allocations: 5.55 KiB)
  78.410 μs (138 allocations: 5.55 KiB)
```
it's between 1-1.5µs on my laptop;
I think this is good to go!